docker/docker_client.go

```diff
@@ -254,6 +254,9 @@ func newDockerClient(sys *types.SystemContext, registry, reference string) (*doc
 		return nil, errors.Wrapf(err, "error loading registries")
 	}
 	if reg != nil {
+		if reg.Blocked {
+			return nil, fmt.Errorf("registry %s is blocked in %s", reg.Prefix, sysregistriesv2.GetConfigPath(sys))
+		}
```

Comment:
Yes, moving this from the various higher-level callers (except for the search path impact, I suppose) is great.

Reply:
That's a good point.

I have no strong opinion on whether search should be allowed. It's currently allowed by Podman, which counts for allowing to search a registry even when being configured as blocked. On the other hand, cutting it off seems cleaner/stricter. Maybe I am over thinking it but if we allow search, would login make sense as well?

@mtrmac @rhatdan @mheon WDYT?

Reply:
I think it breaks down to allowing to opt-out of blocking via the `types.SystemContext`.

Reply:
I think we should block the search if it is blocked in the config. Podman should be fixed.

Reply:
> I have no strong opinion on whether search should be allowed. It's currently allowed by Podman, which counts for allowing to search a registry even when being configured as blocked.

Both Buildah and CRI-O ignore search path elements that are blocked (though with different semantics, Buildah looks up the registry host, CRI-O the result of combining the registry host with the unqualified name).

Ultimately, with this PR, the access is going to fail anyway, so it mostly does not matter whether upper layers do or don’t process the block in the search implementation (except for one rare case, where a search path registry is blocked but the mirror is not: in that case this PR lets the mirror access through, but applying the block at search path level would not allow the mirror to be attempted).

> On the other hand, cutting it off seems cleaner/stricter. Maybe I am over thinking it but if we allow search, would login make sense as well?

I was in no way suggesting that search results should be _exempt_ from blocking (I see no reason for that, let’s keep the features simple and orthogonal if we can); I was just pointing out that there is _some_ code that deals with blocking in some callers of c/image that we might not want to just delete after this PR.

---

As for login, that’s a bit awkward because login is registry-wide but blocking is namespace-wide now. The current behavior (preventing login when the whole registry is blocked, allowing it if only some namespaces are blocked) seems right to me.

---

A kind of opt-out of blocking can be done by setting `SystemRegistriesConfPath` to `/dev/null`. Do we actually need the more detailed granularity? This is supposed to be system-wide policy, it is a bit suspect to provide an easy “let me do it anyway” bit. Of course we can’t _stop_ users from accessing the content (as long as they can compile / execute their own binaries), but I’d wait for an explicit request with a well-reasoned use case before adding that to the stable API.

Reply:
@mtrmac, I interpreted "except for the search path impact, I suppose" as an opt-out. As you said, users could already set it to `/dev/null` and I wouldn't like to add more than that.

> As for login, that’s a bit awkward because login is registry-wide but blocking is namespace-wide now. The current behavior (preventing login when the whole registry is blocked, allowing it if only some namespaces are blocked) seems right to me.

I concur.

> in that case this PR lets the mirror access through, but applying the block at search path level would not allow the mirror to be attempted

I will fix that as I think it would be a bug.

Reply:
Fix which part? OpenShift apparently has some opinions about the intersection of blocking and mirrors… IIRC the idea was to configure mirrors for the primary location _and_ block the primary location, but I’d need to find the earlier conversations or ask again to be sure.

(OTOH the search path aspect is not that relevant for OpenShift.)

Reply:
@mtrmac, to also block here: https://github.com/containers/image/blob/master/docker/docker_image_src.go#L35

I need to test this PR and also need to add new tests for this case to make sure we're not regressing now or later.

> IIRC the idea was to configure mirrors for the primary location and block the primary location

This can already be done with the prefix/location or am I missing something?

> [...] but I’d need to find the earlier conversations or ask again to be sure.

Awesome, thanks!

Reply:
> @mtrmac, to also block here: https://github.com/containers/image/blob/master/docker/docker_image_src.go#L35

Yes, that’s not known to be desirable right now. [Also, as a matter of elegance, having one location to enforce a “blocking” policy is clear; having two risks having three and … Of course, requirements are requirements and elegance is secondary.]

> > IIRC the idea was to configure mirrors for the primary location and block the primary location
> 
> This can already be done with the prefix/location or am I missing something?

That allows _redirecting_ the primary but not disabling its use.

One could convert (Prefix: A, Mirrors: BC) into (Prefix: A, Location: C, Mirrors: B), but that’s pretty nonintuitive. OTOH it might be good enough for OpenShift, where the configuration is going to be machine-generated.

Reply:
> Also, as a matter of elegance, having one location to enforce a “blocking” policy is clear

Yes, that's convincing.

I added integration tests to Skopeo: https://github.com/containers/skopeo/pull/690